File: BeHazeld_OS_Admin/types/catalog.ts
/**
 * Catalog domain types — mirrors FastAPI Pydantic schemas.
 */

export interface CategoryResponse {
  id: string;
  tenant_id: string;
  name: string;
  description: string | null;
  sort_order: number;
}

export interface ProductGroupResponse {
  id: string;
  tenant_id: string;
  name: string;
  description: string | null;
}

export interface ProductTypeResponse {
  id: string;
  tenant_id: string;
  name: string;
}

export interface BrandResponse {
  id: string;
  tenant_id: string;
  name: string;
}

export interface SizeResponse {
  id: string;
  tenant_id: string;
  name: string;
  sort_order: number;
}

export interface ColorResponse {
  id: string;
  tenant_id: string;
  name: string;
  hex_code: string | null;
}

export interface ProductVariantResponse {
  id: string;
  tenant_id: string;
  product_id: string;
  sku_code: string;
  size_id: string;
  color_id: string;
  mrp: string;            // Decimal serialised as string
  selling_price: string;
  cost_price: string | null;
  fabric: string | null;
  reorder_level: string;
  status: string;
}

export interface ProductResponse {
  id: string;
  tenant_id: string;
  product_code: string;
  name: string;
  status: string;
  description: string | null;
  image_url: string | null;
  category_id: string | null;
  product_group_id: string | null;
  product_type_id: string | null;
  brand_id: string | null;
  variants?: ProductVariantResponse[];
}

// ── Request payloads ──────────────────────────────────────────────────────────

export interface CreateProductPayload {
  name: string;
  category_id: string | null;
  product_group_id: string | null;
  product_type_id: string | null;
  brand_id: string | null;
  description: string | null;
  image_url: string | null;
}

export interface CreateVariantPayload {
  size_id: string;
  color_id: string;
  mrp: string;
  selling_price: string;
  cost_price: string;
  fabric: string | null;
  reorder_level: string;
}

export interface CreateMasterPayload {
  name: string;
}

export interface CreateCategoryPayload {
  name: string;
  description: string | null;
  sort_order: number;
}

export interface CreateColorPayload {
  name: string;
  hex_code: string | null;
}

export interface CreateSizePayload {
  name: string;
  sort_order: number;
}

File: BeHazeld_OS_Backend/app/schemas/catalog.py
"""
Catalog Pydantic schemas — request validation and response serialization.

Validation rules
----------------
- name fields: min_length=1 (reject empty strings)
- sort_order / reorder_level: ge=0 (non-negative integers)
- mrp / selling_price / cost_price: gt=0 (strictly positive Decimal)
- hex_code: must match #RRGGBB if provided
- image_url: max_length=500
"""
from __future__ import annotations

import re
import uuid
from datetime import datetime
from decimal import Decimal
from typing import Annotated

from pydantic import BaseModel, ConfigDict, Field, field_validator


# ── Request schemas ───────────────────────────────────────────────────────────

class CreateCategoryRequest(BaseModel):
    name: Annotated[str, Field(min_length=1, max_length=150)]
    description: str | None = None
    sort_order: Annotated[int, Field(ge=0)] = 0


class CreateProductGroupRequest(BaseModel):
    name: Annotated[str, Field(min_length=1, max_length=150)]
    description: str | None = None


class CreateProductTypeRequest(BaseModel):
    name: Annotated[str, Field(min_length=1, max_length=150)]


class CreateBrandRequest(BaseModel):
    name: Annotated[str, Field(min_length=1, max_length=150)]


class CreateSizeRequest(BaseModel):
    name: Annotated[str, Field(min_length=1, max_length=50)]
    sort_order: Annotated[int, Field(ge=0)] = 0


class CreateColorRequest(BaseModel):
    name: Annotated[str, Field(min_length=1, max_length=100)]
    hex_code: str | None = None

    @field_validator("hex_code")
    @classmethod
    def hex_code_must_be_valid(cls, v: str | None) -> str | None:
        if v is not None and not re.match(r"^#[0-9A-Fa-f]{6}$", v):
            raise ValueError("hex_code must be in #RRGGBB format (e.g. '#FF5733')")
        return v


class CreateProductRequest(BaseModel):
    name: Annotated[str, Field(min_length=1, max_length=250)]
    category_id: uuid.UUID | None = None
    product_group_id: uuid.UUID | None = None
    product_type_id: uuid.UUID | None = None
    brand_id: uuid.UUID | None = None
    description: str | None = None
    image_url: Annotated[str | None, Field(max_length=500)] = None


class UpdateProductRequest(CreateProductRequest):
    pass


class CreateVariantRequest(BaseModel):
    size_id: uuid.UUID
    color_id: uuid.UUID
    mrp: Annotated[Decimal, Field(gt=0)]
    selling_price: Annotated[Decimal, Field(gt=0)]
    cost_price: Annotated[Decimal, Field(gt=0)]
    fabric: Annotated[str | None, Field(max_length=150)] = None
    image_url: Annotated[str | None, Field(max_length=500)] = None
    reorder_level: Annotated[int, Field(ge=0)] = 0


class UpdateVariantRequest(CreateVariantRequest):
    pass


class MasterDataImportResponse(BaseModel):
    entity_type: str
    created: int
    skipped: int
    errors: list[str] = []


# ── Response schemas ──────────────────────────────────────────────────────────

class CategoryResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    tenant_id: uuid.UUID
    name: str
    description: str | None
    sort_order: int
    is_active: bool
    created_at: datetime
    updated_at: datetime


class ProductGroupResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    tenant_id: uuid.UUID
    name: str
    description: str | None
    is_active: bool
    created_at: datetime
    updated_at: datetime


class ProductTypeResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    tenant_id: uuid.UUID
    name: str
    is_active: bool
    created_at: datetime
    updated_at: datetime


class BrandResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    tenant_id: uuid.UUID
    name: str
    is_active: bool
    created_at: datetime
    updated_at: datetime


class SizeResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    tenant_id: uuid.UUID
    name: str
    sort_order: int
    is_active: bool
    created_at: datetime
    updated_at: datetime


class ColorResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    tenant_id: uuid.UUID
    name: str
    hex_code: str | None
    is_active: bool
    created_at: datetime
    updated_at: datetime


class ProductVariantResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    tenant_id: uuid.UUID
    product_id: uuid.UUID
    size_id: uuid.UUID
    color_id: uuid.UUID
    sku_code: str
    fabric: str | None
    image_url: str | None
    mrp: Decimal
    selling_price: Decimal
    cost_price: Decimal
    reorder_level: int
    status: str
    created_at: datetime
    updated_at: datetime

    @field_validator("image_url", mode="before")
    @classmethod
    def image_url_must_be_string_or_none(cls, v: object) -> str | None:
        if v is None or isinstance(v, str):
            return v
        return None


class ProductResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    tenant_id: uuid.UUID
    product_code: str
    name: str
    description: str | None
    image_url: str | None
    status: str
    category_id: uuid.UUID | None
    product_group_id: uuid.UUID | None
    product_type_id: uuid.UUID | None
    brand_id: uuid.UUID | None
    created_at: datetime
    updated_at: datetime
    variants: list[ProductVariantResponse] = Field(default_factory=list)

    @field_validator("variants", mode="before")
    @classmethod
    def variants_must_be_list(cls, v: object) -> list[ProductVariantResponse]:
        if isinstance(v, list):
            return v
        return []
